File: src/main/kotlin/me/padej/displayAPI/utils/Segment.java
package me.padej.displayAPI.utils;

import org.bukkit.Location;

public class Segment {
    private final Location pointA;
    private final Location pointB;

    public Segment(Location pointA, Location pointB) {
        this.pointA = pointA;
        this.pointB = pointB;
    }

    public double length() {
        return pointA.distance(pointB);
    }

    public Location getMidpoint() {
        double midX = (pointA.getX() + pointB.getX()) / 2;
        double midY = (pointA.getY() + pointB.getY()) / 2;
        double midZ = (pointA.getZ() + pointB.getZ()) / 2;
        return new Location(pointA.getWorld(), midX, midY, midZ);
    }

    public Location[] getMinMaxPoints() {
        double minX = Math.min(pointA.getX(), pointB.getX());
        double maxX = Math.max(pointA.getX(), pointB.getX());
        double minY = Math.min(pointA.getY(), pointB.getY());
        double maxY = Math.max(pointA.getY(), pointB.getY());
        double minZ = Math.min(pointA.getZ(), pointB.getZ());
        double maxZ = Math.max(pointA.getZ(), pointB.getZ());

        Location minPoint = new Location(pointA.getWorld(), minX, minY, minZ);
        Location maxPoint = new Location(pointA.getWorld(), maxX, maxY, maxZ);

        return new Location[]{minPoint, maxPoint};
    }

    public Location getPointOnSegment(double t) {
        double x = pointA.getX() + (pointB.getX() - pointA.getX()) * t;
        double y = pointA.getY() + (pointB.getY() - pointA.getY()) * t;
        double z = pointA.getZ() + (pointB.getZ() - pointA.getZ()) * t;
        return new Location(pointA.getWorld(), x, y, z);
    }
}

File: src/main/kotlin/me/padej/displayAPI/utils/Segment.kt
package me.padej.displayAPI.utils

import org.bukkit.Location
import kotlin.math.max
import kotlin.math.min

class Segment // Конструктор для создания отрезка с двумя точками
    (// Получить точку A
    private val pointA: Location, // Получить точку B
    private val pointB: Location
) {
    // Метод для вычисления расстояния между точками A и B
    fun length(): Double {
        return pointA.distance(pointB)
    }

    val midpoint: Location
        // Метод для получения координат точек на отрезке
        get() {
            val midX = (pointA.x + pointB.x) / 2
            val midY = (pointA.y + pointB.y) / 2
            val midZ = (pointA.z + pointB.z) / 2
            return Location(pointA.world, midX, midY, midZ)
        }

    val minMaxPoints: Array<Location>
        // Метод для получения минимальной и максимальной точки отрезка
        get() {
            val minX = min(pointA.x, pointB.x)
            val maxX = max(pointA.x, pointB.x)
            val minY = min(pointA.y, pointB.y)
            val maxY = max(pointA.y, pointB.y)
            val minZ = min(pointA.z, pointB.z)
            val maxZ = max(pointA.z, pointB.z)

            val minPoint = Location(pointA.world, minX, minY, minZ)
            val maxPoint = Location(pointA.world, maxX, maxY, maxZ)

            return arrayOf(minPoint, maxPoint)
        }

    // Пример метода для нахождения точки на отрезке по заданному параметру t (где t от 0 до 1)
    // t = 0 — это точка A, t = 1 — это точка B
    fun getPointOnSegment(t: Double): Location {
        val x = pointA.x + (pointB.x - pointA.x) * t
        val y = pointA.y + (pointB.y - pointA.y) * t
        val z = pointA.z + (pointB.z - pointA.z) * t
        return Location(pointA.world, x, y, z)
    }
}



